src/Formats/ProtobufReader.cpp

```diff
@@ -38,26 +38,29 @@ namespace
     // Those inequations helps checking conditions in ProtobufReader::SimpleReader.
     constexpr Int64 END_OF_VARINT = -1;
     constexpr Int64 END_OF_GROUP = -2;
+    constexpr Int64 END_OF_FILE = -3;
 
     Int64 decodeZigZag(UInt64 n) { return static_cast<Int64>((n >> 1) ^ (~(n & 1) + 1)); }
 
-    [[noreturn]] void throwUnknownFormat()
-    {
-        throw Exception("Protobuf messages are corrupted or don't match the provided schema. Please note that Protobuf stream is length-delimited: every message is prefixed by its length in varint.", ErrorCodes::UNKNOWN_PROTOBUF_FORMAT);
-    }
 }
 
 
 // SimpleReader is an utility class to deserialize protobufs.
 // Knows nothing about protobuf schemas, just provides useful functions to deserialize data.
-ProtobufReader::SimpleReader::SimpleReader(ReadBuffer & in_)
+ProtobufReader::SimpleReader::SimpleReader(ReadBuffer & in_, const bool single_message_mode_)
```

Comment:
I think `single_message_mode_` should passed as an argument to `SimpleReader::startMessage()`, not to the constructor.

Reply:
And what are the advantages? 

Then it will need to be stored inside `ProtobufReader::ProtobufReader` or inside `ProtobufRowInputFormat::ProtobufRowInputFormat` (now they just pass that down). 

Now once the simple writer is consructed all the logic or writing is inside. Your option suggests we can write one message with header and another without using single writer / reader

Reply:
You have to store this flag inside `SimpleReader` anyway. The advantage is that it makes the code clearer because so you pass to the function `startMessage()` the parameters of reading the message that you're reading.

> Your option suggests we can write one message with header and another without using same writer / reader by swiching the flags during start / end message (more error-prone)

Yes, but that's ok because the class `SimpleReader` is an internal class, so it's the caller's responsibility to use it correctly.